New mapgen and gunstores

mugling opened the pull request:
Continues from #15318 and implements new mapgen section `place_loot` which provides:
- absolute percentage `chance` to spawn `repeat` iterations of `group` or `item`
- percentage chance for each item to spawn with a `magazine`
- percentage chance for each item to spawn with `ammo`

`magazine` and `ammo` can be specified independently of each other so you for example have a 70% chance for an item to contain a magazine, but only 30% for it to contain ammo.

This PR provides the minimum implementation to re-implement gunstore spawns using the itemgroups from #15275, #15319 and others. I'm happy extend `place_loot` if there are any suggestions for extra functionality but that would be a separate PR.

The intended balance for the gunstores is as follows:

**Regular gunstore**
Arrive and leave with a working gun.
- Contains common pistols, rifles and shotguns and to a lesser extent SMGs
- Infrequently spawns some rarer guns
- Enough common ammo to reload a few magazines, all types but biased towards pistol
- Stocks a few common magazines and the occasional rare one
- Range at back usually has one (often loaded) pistol, sometimes a rare or obscure variant
- Level 1 skill books are often found on the counter

**Defense gunstore**
Stock up on ammo and magazines for your existing guns
- Much fewer guns but a lot more choice as to ammo and magazines
- Lots of common ammo including for rifles and shotguns
- Large stock of magazines including the rarer variants
- Lockers at back of store contain rare guns (usually with magazines) and ammo

**Gunsmith**
Adjust your existing gun or reload ammo for it
- No guns except for infrequent spawns of the obscure types in the shop window
- No ammo but plentiful supplies for reloading it in the workshop
- Workshop guaranteed to contain a hand press and kinetic puller plus other tooling
- Most (presently all) gunmods plus a selection of the rarer magazines behind the counter
- Firearm and gunsmith repair kits can often be found in the shop window
- One or more higher level skill books

Variety of spawned guns will automatically improve as more are migrated to the new format so can be merged prior to this ~~however presently marked [WIP] whilst I finish off the gunsmith store~~

mugling commented:
As of a5ab6b4 `place_loot` now support s placing by item `id` as well as `item_group`

illi-kun commented on data/json/itemgroups/books.json:
```diff
@@ -5,10 +5,25 @@
+  {
+    "type" : "item_group",
+    "id" : "book_gunref",
+    "//" : "Reference boooks about guns",
```
Nitpicking: are these triple 'o' intended?

tyrael93 commented:
Very interesting! A little side note, what about those basements that spawn with lots of guns and weapons in them? I remember finding some rivtech stuff and flamethrowers in there, not to mention katanas, but that's just an OP weapon not a firearm.

tyrael93 commented:
Oh and almost forgot. NPCs spawn with old ammo. They cannot use their own guns and start bashing monsters in the head with it. Saw a dude with an assault rifle just start to beat the shit out if a soldier ant and when I found his corpse he was full of old ammo for his gun.

Just throwing it out there. Thanks!

BevapDin commented on src/mapgen.cpp:
```diff
@@ -10620,6 +10657,64 @@ int map::place_items(items_location loc, int chance, int x1, int y1,
     return item_num;
 }
 
+int map::place_loot( std::string id, int chance, int x1, int y1, int x2, int y2, int ammo, int magazine, int turn )
```
It's a bit sad that one needs to pass the id of different things (item groups vs item type) in the same parameter. It defeats the whole purpose of having different types.

BevapDin commented on src/mapgen.cpp:
```diff
@@ -830,6 +832,40 @@ class jmapgen_item_group : public jmapgen_piece {
+        }
+        void apply( map &m, const jmapgen_int &x, const jmapgen_int &y, const float /*mon_density*/ ) const override
+        {
+            m.place_loot( group.empty() ? name : group, chance.get(), x.val, y.val, x.valmax, y.valmax, ammo.get(), magazine.get() );
```
What happens when the JSON contains both, an item id *and* a group id?

mugling commented:
> A little side note, what about those basements

Convert them in to JSON then provide some variants

mugling commented:
@BevapDin I've moved the functionality inline as ~40% of the function was duplicated error checks in any case. Also added in cc2f407 is a check that one (and only one) of `item` or `group` is set.
